src/tests/test_serializer.py

```diff
@@ -222,3 +230,72 @@ def test_deserialize_complex(self):
         loaded_obj = json_loads(serialized_obj)
         assert type(loaded_obj) == complex
         assert loaded_obj == test_obj
+
+
+@pytest.mark.parametrize("backend", [SerialiserBackend.json, SerialiserBackend.ujson])
+class TestBuilderSerialiser:
+    """
+    Note here, that typically doing a direct float comparison is not meaningful in a test.
+    Here however, we need to ensure that the integrity of the floats is exactly preserved across serialisation.
+    """
+
+    floating_point_num = math.sqrt(math.pi) / 2.14324354
+
+    @pytest.fixture
+    def builder(self):
+        hardware = get_default_echo_hardware(4)
+
+        hardware.get_qubit(0).get_drive_channel().frequency = self.floating_point_num
+
+        builder = get_builder(hardware)
+        builder.X(hardware.get_qubit(0))
+        builder.U(
+            hardware.get_qubit(0),
+            theta=self.floating_point_num,
+            phi=self.floating_point_num,
+            lamb=self.floating_point_num,
+        )
+        return builder
+
+    def test_floating_point_numbers_hardware(self, backend, builder):
+        """
+        Testing that different serialisation backends respect floating point numbers in hardware
+        """
+        string = builder.serialize(backend=backend)
+
+        new_builder = InstructionBuilder.deserialize(string, backend=backend)
+
+        assert (
+            new_builder.model.get_qubit(0).get_drive_channel().frequency
+            == self.floating_point_num
+        )
+
+    def test_floating_point_numbers_instructions(self, backend, builder):
+        """
+        Testing that different serialisation backends respect floating point numbers in hardware
+
+        """
+        string = builder.serialize(backend=backend)
+
+        new_builder = InstructionBuilder.deserialize(string, backend=backend)
+
+        for instruction in new_builder.instructions:
+            if isinstance(instruction, PhaseShift):
+                phase = instruction.phase
+                if isclose(phase, self.floating_point_num):
+                    # There are phase instructions that aren't just the pure rotation, easiest way to find them
+                    assert instruction.phase == self.floating_point_num
```

Comment:
This is a rare exception to using a == operator. However it might be more interesting to convert to a bit sequence and compare that. At least you can then go off and debug what the serialisation library has done. For example https://pypi.org/project/bitstring/

Reply:
I don't quite understand the value of this. We're only seeing the effect of the final result in python so would see that a bit string is different or a float is different. This is effectively a backstop test of a third party library that I'd usually not both with, but just need this for integrity checking